### PR TITLE
Fix CFileTree::GetNodeFullPath memory leak

### PR DESCRIPTION
**Pull request by Toilal**
It also use const char* instead of char* for immutable strings

### CODE REVIEW

File: src/FileTree.cpp [hunk 1]
```diff
@@ -265,11 +265,11 @@ tree_node_<FILE_ITEM>* CFileTree::findParentNodeFromRootForPath(char *path) {
 	if (followingPath.empty()) {
 		return topNode.node;
 	} else {
-		return findNodeWithPathFromNode(followingPath, topNode.node);
+		return findNodeWithPathFromNode(followingPath.c_str(), topNode.node);
 	}
 }
 
-char* CFileTree::GetNodeFullPath(tree_node_<FILE_ITEM>* node)
+std::string CFileTree::GetNodeFullPath(tree_node_<FILE_ITEM>* node)
```
Comment (Toilal) [inline src/FileTree.cpp]: The effective fix is here (returning std::string instead of char*).

**Comment by Toilal** [comment]
After more that one day running this fork, #52 didn't occur. The process actually eats up far less memory than it did (actually 52 016K for Private Bytes, and 2 776K for Working Set).

**Comment by Toilal** [comment]
Here's the compiled EXE for people who wants to test this forks. It includes #53 and #51.

[WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/1351251/WinNFSd.zip)

**Comment by marcharding** [comment]
Hey, looks good. https://github.com/winnfsd/winnfsd/issues/34 had the same approach but i had problems after i applied the patch.

(I realized this week that this was probably due to the introduced canonical path functions in the https://github.com/winnfsd/winnfsd/issues/34 patch above, GetFullPathName/GetFullPathNameW would solve that)

I'll testrun your fixes the next few days 👍

**Comment by Toilal** [comment]
Sadly it doesn't seem to fix #29 as I hit this exact issue this morning, but it really seems to fix #52 and reduce memory usage drastically.

**Comment by Toilal** [comment]
I still need to work on this. 

In some case, getPathByHandle can return NULL, but it cause a crash because method has to return std::string. I'll add an exception instead of returning NULL, and catch it to return a proper error to NFS client.

**Comment by Toilal** [comment]
I'll open another pull request because I don't think introducing std::string was a so good idea ... I think it's better to keep using a char pointer to avoid creating to much std::string object in the call stack. Stay tuned

**Comment by Toilal** [comment]
@marcharding, I just add a new commit bcf16896d11db47299e596b3ed3cd161a2f2df74 that fixes crashes when an invalid handle (unknown tree item) is requested by the client.

It has occured on my environment in a reproductible manner, so I could debug and fix this issue properly. It also pass std::string by reference. That's the first time I write C++, so I was a bit confused at first with those changes, but I think I now understand the keys of memory management.

I attach the binary from HEAD of this branch for people who wants to test.

[WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/1379743/WinNFSd.zip)

**Comment by marcharding** [comment]
Looks good so far, i'll test it a few more days to see if there are any regressions. 👍 👍 👍

**Comment by kriks57** [comment]
I've been using it for some days now and it's stable for me. The memory consumption stopped to increase and is stabilized around 50MB.
thanks!

**Comment by Toilal** [comment]
We have been running this on a team of 5 developers since one month, and no more crash too, at least when using UDP protocol.

I think it could be merged and released @marcharding please :)

**Comment by marcharding** [comment]
Will do later today 👍

**Pull request merged**